### PR TITLE
Don't phi-merge uninitialised objects

### PR DESCRIPTION
**Pull request by JohnDumbell**
This is the work mentioned in the last comment on #2121, making it so that a generation 0 of a symex variable means it's never been initialised in both Java and C. The approach is on an allocate to assign a nondet to it if it isn't zero-init'd to begin with, and then if we see a gen zero then to not consider it during conditionals on phi merges.

Along with these is a small change to test.pl's multi-line match option to enable traditional regex multi-line matching. The m flag causes ^ and $ to actually match correctly on new line/carriage returns, where before the match was actually on the entire file so ^ was just beginning of file $ was the end. This also fixes a few tests' failure condition where this subtle difference wasn't noticed and '^warning: ignoring' is used.

### CODE REVIEW

File: regression/test.pl [hunk 1]
```diff
@@ -174,7 +174,7 @@ ($$$$$$$$$)
             binmode $fh;
             my $whole_file = <$fh>;
             $whole_file =~ s/\r\n/\n/g;
-            my $is_match = $whole_file =~ /$result/;
+            my $is_match = $whole_file =~ /$result/m;
```
Comment (tautschnig) [inline regression/test.pl:177]: This is great, but could it please be put in a separate commit? The explanation in the PR is absolutely necessary to understand this, but will never make it to the commit history at this stage.

**Review by tautschnig — CHANGES_REQUESTED** [review]
~I'm struggling to understand why this proposal is the adequate solution for the problem. From #2121 it seems the problem is one of the points-to analysis not being sufficiently smart. I'll comment more over there.~ (Edit: That was me being dense.)

With this PR the commit message doesn't have information that would enable a future reader of this commit to understand what's going on. Please make a habit of writing  good commit messages.

Edit 2: We also need tests outside Java.

File: regression/cbmc-java/PhiMergeConstantPropogation/Main.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
+public class Main {
+
+    public int phiMerge(Boolean trigger) {
```
Comment (smowton) [inline regression/cbmc-java/PhiMergeConstantPropogation/Main.java]: sp s/propogation/propagation (in filename and elsewhere)

File: regression/cbmc-java/PhiMergeConstantPropogation/test.desc [hunk 3]
```diff
@@ -0,0 +1,7 @@
+CORE 
+Main.class
+--function Main.phiMerge --show-vcc
+activate-multi-line-match
+^.*dynamic_object2#2 == \{ \.@class_identifier="java::Main\$Temp",\s*\.@lock=false \}$
```
Comment (smowton) [inline regression/cbmc-java/PhiMergeConstantPropogation/test.desc]: Should probably check for the other dynamic object too (the instance of `Temp` and the instance of `Ephemeral`)

File: src/goto-symex/symex_goto.cpp [hunk 4]
```diff
@@ -429,6 +429,14 @@ void goto_symext::phi_function(
       rhs=goto_state_rhs;
     else if(goto_state.guard.is_false())
       rhs=dest_state_rhs;
+    else if(goto_state.level2_current_count(l1_identifier) == 0)
```
Comment (smowton) [inline src/goto-symex/symex_goto.cpp:436]: Add comments briefly explaining why these checks are here

File: src/goto-symex/symex_builtin_functions.cpp [hunk 5]
```diff
@@ -184,6 +184,12 @@ void goto_symext::symex_allocate(
     else
       throw "failed to zero initialize dynamic object";
   }
+  else
+  {
+    exprt nondet = build_symex_nondet(object_type);
```
Comment (smowton) [inline src/goto-symex/symex_builtin_functions.cpp:189]: Point of interest: you've copied the case for `x = NONDET(int)` rather than the case from `code_declt("x", java_int_type())` for example. Check how their behaviour differs if at all.

**Review by smowton — COMMENTED** [review]
Looks good, some small changes

Additional tests that would be good (in each case for both C and Java): check that locals, static locals (namespaced globals), normal globals and dynamic allocations all behave as expected.

File: src/goto-symex/symex_builtin_functions.cpp (hunk 5 above)
Comment (JohnDumbell) [reply src/goto-symex/symex_builtin_functions.cpp:189]: I can't see any major deviations - the code isn't identical but the intentions seem the same and most of the code in decl is run in assign at some point too. One thing to note is that assign propagates the nondet as a constant, while the decl code flat-out disables any propagation, would that be an issue?

Comment (smowton) [reply src/goto-symex/symex_builtin_functions.cpp:189]: It'll only disable it for the time being, but it's interesting. I'm not quite well informed enough to give an authoritative answer -- I suggest running this by @peterschrammel and post his answer here.

File: regression/cbmc-java/PhiMergeConstantPropogation/test.desc (hunk 3 above)
Comment (JohnDumbell) [reply regression/cbmc-java/PhiMergeConstantPropogation/test.desc]: Decided to change the logic to make it more robust, it now just makes sure there's no gen zero objects in guards, doesn't matter where they are.

**Comment by JohnDumbell** [comment]
Added a new set of tests and made some minor revisions to the code and commits in regards to comments (sans talking to peter). I'm not well versed in C so if there's a potential condition I've missed just say.

File: regression/test.pl (hunk 1 above)
Comment (tautschnig) [reply regression/test.pl:177]: Thanks for factoring it out into a separate commit. Could you please also add the text from the PR to body of the commit message? It's really informative, and should be preserved!

File: src/goto-symex/symex_builtin_functions.cpp (hunk 5 above)
Comment (tautschnig) [reply src/goto-symex/symex_builtin_functions.cpp:189]: I'd argue that this is what #35 is about, and you're doing parts of that here, even though the full value only unfolds once you break down nondet symbols into their individual fields.

**Review by tautschnig — APPROVED** [review]
As expressed in a few places by reviewers: more comments and documentation (commit messages) needed, but otherwise looks good to me.

File: regression/test.pl (hunk 1 above)
Comment (JohnDumbell) [reply regression/test.pl:177]: Ah, sorry, fixed. Still getting fully used to git.

File: src/goto-symex/symex_goto.cpp (hunk 4 above)
Comment (JohnDumbell) [reply src/goto-symex/symex_goto.cpp:436]: Added a comment block at the top of the if.

File: regression/cbmc-java/phi-merge_uninitialized_values/PhiMergeUninitialized.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            obj = new Ephemeral(42);
+        } else {
+            obj = new Aetherial(20);
+        }
+
+        assert obj.val == 20;
+        return obj.val;
+    }
+
+    private Ephemeral local;
```
Comment (smowton) [inline regression/cbmc-java/phi-merge_uninitialized_values/PhiMergeUninitialized.java]: This isn't really a local -- in Java we can't really have a local object at all (but we can have a local reference or pointer). Really the one labelled `dynamic` is using a local variable to store the pointer, and this one is using a member field to store it.

File: regression/cbmc-java/phi-merge_uninitialized_values/PhiMergeUninitialized.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            local = new Ephemeral(42);
+        } else {
+            local = new Aetherial(50);
+        }
+
+        assert local.val == 42;
+        return local.val;
+    }
+
+    private static Ephemeral staticLocal;
```
Comment (smowton) [inline regression/cbmc-java/phi-merge_uninitialized_values/PhiMergeUninitialized.java]: Again, `staticField` would be more accurate

File: regression/cbmc-java/phi-merge_uninitialized_values/local.desc [hunk 8]
```diff
@@ -0,0 +1,9 @@
+CORE 
+PhiMergeUninitialized.class
+--function PhiMergeUninitialized.dynamicAllocationUninitialized --show-vcc
+activate-multi-line-match
+^EXIT=0$
+^SIGNAL=0$
+--
+^.*: (dynamic_object|new_tmp[0-9]+(@[0-9]+)?)[0-9]+#0\)
+^.*\? (dynamic_object|new_tmp[0-9]+(@[0-9]+)?)[0-9]+#0
```
Comment (smowton) [inline regression/cbmc-java/phi-merge_uninitialized_values/local.desc]: In test.desc files a second `--` delimits comments. You should add one and then comment on what these regexes are looking for.

Format example:

```
CORE
Main.class
--other --arguments
Must find this
Also must find this
--
Musn't find this
Nor this
--
This is a comment
This is another comment
```

File: regression/cbmc/phi-merge_uninitialized_values/test.c [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  }
+  else
+  {
+    *obj = 20;
+  }
+
+  assert(*obj == 20);
+}
+
+int *staticLocal;
```
Comment (smowton) [inline regression/cbmc/phi-merge_uninitialized_values/test.c]: In C we really can have local objects. For all the cases other than `dynamic`, don't use malloc, instead declare the object on the stack, for example `int local;`

Static locals are possible in C, and are effectively globals that are only accessible in that function. For example,
```c
int f() {
  static int staticvar = 100;
  return staticvar++;
}

void main() {
  assert(f() == 100); // Passes
  assert(f() == 101); // Passes
}
```

The final missing case for C is a "real" global. For example:
```c
int global;

int main() {
  global = 1234;
  ...
}
```

File: src/goto-symex/symex_goto.cpp (hunk 4 above)
Comment (tautschnig) [reply src/goto-symex/symex_goto.cpp:436]: Thank you!

File: regression/cbmc/phi-merge_uninitialized_values/test.c [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+{
+  int *obj;
+  obj = malloc(sizeof(int));
+  if(trigger)
+  {
+    *obj = 42;
+  }
+  else
+  {
+    *obj = 20;
```
Comment (peterschrammel) [inline regression/cbmc/phi-merge_uninitialized_values/test.c]: Maybe also add a test where one branch is missing and the merge must happen.

File: regression/cbmc/phi-merge_uninitialized_values/test.c [hunk 11]
```diff
@@ -0,0 +1,55 @@
+#include <assert.h>
+#include <stdlib.h>
+
+void dynamicAllocationUninitialized(int trigger)
+{
+  int *obj;
+  obj = (int*)malloc(sizeof(int));
```
Comment (smowton) [inline regression/cbmc/phi-merge_uninitialized_values/test.c]: For symmetry with the Java case this should only be allocated on one branch (i.e. we should allocate within the `if` / `else`)

**Comment by smowton** [comment]
This is looking much improved, just commented with one more nitpick, let me know when this is ready for re-review

**Review by smowton — APPROVED** [review]

**Comment by smowton** [comment]
Looks great, merge once tests are passing I say

**Comment by JohnDumbell** [comment]
Made all requested changes, I'll fix any issues that the CI flags up but right now it looks like some weird timeout issues, so I'll just re-run it until it gives me something useful or passes.

**Comment by tautschnig** [comment]
https://www.traviscistatus.com/ says there are problems with macOS, so it might take a little longer... (all other jobs have passed here).
